### ISSUE TITLE
Scrolling will be immediately cancelled when scrolling is *triggered* by keyDown events

### ISSUE THREAD

**Issue by sleexyz** [fisshy/react-scroll#269]
I understand that by default, `keyDown` events will cancel scrolls, and that that behavior is cancellable by setting `ignoreCancelEvents` to true.

However, a bug we're running into right now is that when scrolling is *triggered* by a React `keyDown` handler, the scrolling is immediately cancelled. This is errorneous behavior; the scroll cancellation should only be able to kick in *after* the scroll event is triggered. In other words, there is no way to synchronously trigger scrolls from a `keyDown` event handler without using `ignoreCancelEvents`.

The reason why the scroll immediately gets cancelled is that the event listeners on `document` are added *after* the element tree is mounted to the DOM, i.e. after `ReactDOM.render` is called.

https://github.com/fisshy/react-scroll/blob/886a792b10e4c72731c3652c5922583e165b4798/modules/mixins/passive-event-listeners.js#L19

When `document` event listeners are added after `ReactDOM.render` is called, then those event listeners will fire *after* react's `document` event listeners fire.

[(Demonstration of document.addEventListener behavior after ReactDOM render)](https://codesandbox.io/s/y3n5p5oxnz)

So what happens is
1. a key is pressed
2. React handles the keyDown event
3. `ReactScroll.scroller.scrollTo` is called, initializing the scrolling animation
4. ReactScroll handles the keyDown event, and cancels the scroll

---

I think a proper implementation of scroll cancelling would bind the event listeners on `document` only once,  during module initialization. This ensures that its event handlers on `document` are bound before ReactDOM's. This solution would also solve the memory leak mentioned in this issue: #268.

A workaround is to set `ignoreCancelEvents` to true.

**Comment by fisshy** [comment]
So, removing events after scroll would also work right?

**Comment by fisshy** [comment]
I haven't published any new version, but this one works for me.

check out the latest commit.

https://github.com/fisshy/react-scroll/commit/eda9f2b9dac503b6741fcbf0619a12c5f74517c1

`npm run examples`

**Comment by sleexyz** [comment]
Tried it out the new keyDown example, still partially broken. If you keyDown with a particular cadence, then only every other scroll will work. The others will be cancelled.

I believe the crux of the issue is that we're registering event handlers on `document` after ReactDOM.render is called. We should register a "mother" event handler on `document` once during module initialization time, and then on each call to `animateTopScroll`, we should register (and subsequently deregister) child event handlers onto that "mother" event handler. This way we can ensure that the cancellation will occur strictly before render is finished.

**Comment by fisshy** [comment]
Interesting. How would the "mother event" tell the scroller to stop scroll?

I'm not sure if "key down" after a "key down" is broken because it's part of the cancellation events. 

Another approach could be to specify what cancel events you want, ignoring keydown would solve it.

**Comment by sleexyz** [comment]
I realized the `__cancel` variable is a top level variable anyways.

https://github.com/fisshy/react-scroll/blob/eda9f2b9dac503b6741fcbf0619a12c5f74517c1/modules/mixins/animate-scroll.js#L46

So the "mother" event handler doesn't need to be as complicated as I imagined it would need to me. Something like:

```
let __cancel = false;
document.addEventListener('keydown', () => {
  __cancel = true;
});
```
So instead of registering event listeners every call to `animateTopScroll`, we only register one event handler once, in the beginning when the module is loaded.
